hal/src/photon/concurrent_hal.cpp

```diff
@@ -114,11 +103,239 @@ os_result_t os_thread_join(os_thread_t* thread)
  * @param thread    The thread to clean up.
  * @return 0 on success.
  */
-os_result_t os_thread_cleanup(os_thread_t* thread)
+os_result_t os_thread_cleanup(os_thread_t thread)
 {
-    return wiced_rtos_delete_thread((wiced_thread_t*)thread);
+    vTaskDelete(thread);
+    return 0;
 }
 
+/**
+ * Map gthread handles to FreeRTOS task handles.
+ */
+typedef TaskHandle_t __gthread_t;
+
+
+class ThreadQueue
+{
+    QueueHandle_t queue;
+
+public:
+
+    ThreadQueue(UBaseType_t max_size) {
+        queue = xQueueCreate(max_size, sizeof(TaskHandle_t));
+    }
+
+    ~ThreadQueue() {
+        vQueueDelete(queue);
+    }
+
+
+    void enqueue(TicksType_t ticks=portMAX_DELAY) {
+        TaskHandle_t current = xTaskGetCurrentTaskHandle();
+        xQueueSend(queue, &current, ticks);
+    }
+
+    void sleep() {
+        enqueue();
+        vTaskSuspend(NULL);
+    }
+
+    bool wake() {
+        TaskHandle_t next;
+        if (xQueueReceive(queue, &next, 0)==pdTRUE) {
+            vTaskResume(next);
+            return true;
+        }
+        return false;
+    }
+
+    void wakeAll()
+    {
+        while (wake()) {}
+    }
+};
+
+
+class Lock
+{
+    SemaphoreHandle_t semaphore;
+
+public:
+
+    Lock() {
+        semaphore = xSemaphoreCreateMutex();
+    }
+
+    void acquire() {
+        while (xSemaphoreTake(semaphore, portMAX_DELAY)==pdFALSE);
+    }
+
+    void release() {
+        xSemaphoreGive(semaphore);
+    }
+
+    ~Lock() {
+        vSemaphoreDelete(semaphore);
+    }
+};
+
+
+class ConditionVariable
+{
+    ThreadQueue queue;
+
+    typedef std::unique_lock<std::mutex> lock_t;
+public:
+    ConditionVariable(UBaseType_t max_size) : queue(max_size) {}
+
+    void wait(lock_t* lock)
+    {
+        taskENTER_CRITICAL();
+        lock->unlock();
+        queue.enqueue();
+        taskEXIT_CRITICAL();
+        vTaskSuspend(NULL);
+        lock->lock();
+    }
```

Comment:
The scheduler preempting a task is not the same at suspending it.

The vTaskSuspend docs say "When suspended a task will never get any microcontroller processing time, no matter what its priority"

The vTaskResume docs say "A task that has been suspended by one or more calls to vTaskSuspend () will be made available for running"

So in case vTaskResume is called first on the task trying to wait it will be a no-op since it has not been suspended yet (not currently running, but allowed to be run by the scheduler). Then when the task trying to wait is scheduled again it will call vTaskSuspend and be blocked from executing forever.